File: 652_Find_Duplicate_Subtrees/652_1.py
from collections import defaultdict
from typing import List

class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def findDuplicateSubtrees(self, root: TreeNode) -> List[TreeNode]:
        res = []
        trees = defaultdict(lambda: 0)

        def findSubTree(node):
            if not node:
                return "#"
            left =  findSubTree(node.left)
            right = findSubTree(node.right)
            subtree = str(node.val) + "," + left + "," + right
            if trees[subtree] == 1:
                res.append(node)
            trees[subtree] += 1
            return subtree

        findSubTree(root)
        return res

File: 652_Find_Duplicate_Subtrees/652_1.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    vector<TreeNode*> findDuplicateSubtrees(TreeNode* root) {
        vector<TreeNode*> res;
        unordered_map<string, int> map;
        findSubtree(root, res, map);
        return res;
    }

    string findSubtree(TreeNode *node, vector<TreeNode*> &res, 
                       unordered_map<string, int> &map) {
        if (!node) return "#";
        string left =  findSubtree(node->left, res, map);
        string right = findSubtree(node->right, res, map);
        string subTree = to_string(node->val) + "," + left + "," + right;
        if (map[subTree] == 1) res.emplace_back(node);
        map[subTree]++;
        return subTree;
    }
};
